src/plugins_types_bits.c

```diff
@@ -62,113 +62,120 @@ ly_type_store_bits(const struct ly_ctx *ctx, const struct lysc_type *type, const
     ret = ly_set_new(&items);
     LY_CHECK_GOTO(ret != LY_SUCCESS, cleanup_value);
 
-    ws_count = lws_count = 0;
-    for (index = 0; index < value_len; ++index) {
-        if (isspace(value[index])) {
-            ++ws_count;
-            continue;
+    /* get all values */
+    while (index_end < value_len) {
+        /* skip leading spaces */
+        index = index_end;
+        while ((index < value_len) && isspace(value[index])) {
+            index++;
         }
-        if (index == ws_count) {
-            lws_count = ws_count;
-        } else if (ws_count) {
-            iscanonical = 0;
+
+        index_end = index;
+        /* find end of word */
+        while ((index_end < value_len) && !isspace(value[index_end])) {
+            index_end++;
         }
-        ws_count = 0;
 
-        /* start of the item */
+        /* check if name of bit is valid */
         item = &value[index];
-        for (item_len = 0; index + item_len < value_len && !isspace(item[item_len]); item_len++) {}
-        LY_ARRAY_FOR(type_bits->bits, u) {
-            if (!ly_strncmp(type_bits->bits[u].name, item, item_len)) {
-                /* we have the match */
-                uint32_t inserted;
-
-                if (iscanonical && items->count && (type_bits->bits[u].position < ((struct lysc_type_bitenum_item *)items->objs[items->count - 1])->position)) {
-                    iscanonical = 0;
-                }
-                ret = ly_set_add(items, &type_bits->bits[u], 0, &inserted);
-                LY_CHECK_GOTO(ret, cleanup);
-                if (inserted != items->count - 1) {
-                    ret = ly_err_new(err, LY_EVALID, LYVE_DATA, NULL, NULL,
-                            "Bit \"%s\" used multiple times.", type_bits->bits[u].name);
-                    goto cleanup;
-                }
-                goto next;
+        item_len = index_end - index;
+        if (item_len != 0) {
```

Comment:
your code doesn't cover `item_len == 0`

Reply:
ok, ignore the previous note, but maybe it would be clearer to detect earlier that the while loop was terminated because of `index < value_len` condition  (so `index_len` and `index` are the same) and the whole loop can be terminated (no item to store). Then this condition here will not be needed.

Reply:
I don't think so. The string can contain only white characters then we need to skip this character.  Probably I should change variable name _index_ to _index_start_.

Reply:
yes, but I'm talking about the clarity of the code. What you describe is actually detection that there is no (more) bit name in the string (so not just when the string contains only spaces, but you will end up here also with the trailing spaces). Try just to invert the condition: if `item_len == 0` then terminate the loop with the appropriate comment. The rest of the bit name processing inside this `if` block can be then just put after that loop termination condition even without indentation.

Reply:
Then I would have to use command goto on non-error handling. OK, I'm going to do it but I really don't like it. As I write before. I'm convinced that command goto should be used only for error handling.

Reply:
I believe it can be done without the label and goto - see the comment in the updated code.